File: frontend/src/landing/app/mocks/matchDetail.ts
import type { MatchDetail, ParticipantDetail, TeamDetail } from "../types/match";

const VIEWER_PUUID = "mock-viewer-puuid";

function participant(
  overrides: Partial<ParticipantDetail> & Pick<ParticipantDetail, "champion_name" | "champion_id" | "position" | "team_win">
): ParticipantDetail {
  const win = overrides.team_win;
  return {
    puuid: overrides.puuid ?? `puuid-${overrides.champion_name}`,
    riot_id: overrides.riot_id ?? `${overrides.champion_name}Player#EUW`,
    champion_id: overrides.champion_id,
    champion_name: overrides.champion_name,
    position: overrides.position,
    win,
    kills: overrides.kills ?? 3,
    deaths: overrides.deaths ?? 4,
    assists: overrides.assists ?? 8,
    cs: overrides.cs ?? 180,
    gold_earned: overrides.gold_earned ?? 11_000,
    damage_to_champions: overrides.damage_to_champions ?? 18_000,
    vision_score: overrides.vision_score ?? 25,
    items: overrides.items ?? [3031, 3006, 3046, 3035, 3036, 0, 3363],
    summoner_spells: overrides.summoner_spells ?? [4, 14],
    is_viewer: overrides.is_viewer ?? false,
  };
}

function buildTeams(viewerWin: boolean): readonly [TeamDetail, TeamDetail] {
  const blueWin = viewerWin;
  const redWin = !viewerWin;

  const blue: TeamDetail = {
    team_id: 100,
    win: blueWin,
    bans: [157, 64, 238, 555, 89],
    objectives: {
      baron: blueWin ? 1 : 0,
      dragon: 3,
      rift_herald: 1,
      tower: 7,
      inhibitor: blueWin ? 2 : 0,
    },
    participants: [
      participant({
        champion_name: "Garen",
        champion_id: 86,
        position: "TOP",
        team_win: blueWin,
        kills: 5,
        deaths: 2,
        assists: 4,
        cs: 198,
      }),
      participant({
        champion_name: "LeeSin",
        champion_id: 64,
        position: "JUNGLE",
        team_win: blueWin,
        kills: 4,
        deaths: 5,
        assists: 12,
        cs: 142,
        summoner_spells: [4, 11],
      }),
      participant({
        champion_name: "Ahri",
        champion_id: 103,
        position: "MIDDLE",
        team_win: blueWin,
        kills: 8,
        deaths: 3,
        assists: 6,
        cs: 210,
      }),
      participant({
        champion_name: "Jinx",
        champion_id: 222,
        position: "BOTTOM",
        team_win: blueWin,
        kills: 12,
        deaths: 4,
        assists: 5,
        cs: 245,
        is_viewer: true,
        puuid: VIEWER_PUUID,
        riot_id: "You#EUW",
      }),
      participant({
        champion_name: "Thresh",
        champion_id: 412,
        position: "UTILITY",
        team_win: blueWin,
        kills: 1,
        deaths: 6,
        assists: 18,
        cs: 32,
        vision_score: 78,
      }),
    ],
  };

  const red: TeamDetail = {
    team_id: 200,
    win: redWin,
    bans: [222, 51, 254, 777, 360],
    objectives: {
      baron: redWin ? 1 : 0,
      dragon: 2,
      rift_herald: 0,
      tower: 4,
      inhibitor: redWin ? 1 : 0,
    },
    participants: [
      participant({
        champion_name: "Darius",
        champion_id: 122,
        position: "TOP",
        team_win: redWin,
        kills: 6,
        deaths: 4,
        assists: 2,
      }),
      participant({
        champion_name: "Viego",
        champion_id: 234,
        position: "JUNGLE",
        team_win: redWin,
        kills: 7,
        deaths: 6,
        assists: 4,
        summoner_spells: [4, 11],
      }),
      participant({
        champion_name: "Syndra",
        champion_id: 134,
        position: "MIDDLE",
        team_win: redWin,
        kills: 5,
        deaths: 5,
        assists: 7,
      }),
      participant({
        champion_name: "Caitlyn",
        champion_id: 51,
        position: "BOTTOM",
        team_win: redWin,
        kills: 9,
        deaths: 5,
        assists: 3,
      }),
      participant({
        champion_name: "Lux",
        champion_id: 99,
        position: "UTILITY",
        team_win: redWin,
        kills: 2,
        deaths: 7,
        assists: 14,
        vision_score: 65,
      }),
    ],
  };

  return [blue, red];
}

function createMatch(
  matchId: string,
  viewerWin: boolean,
  gameDuration: number
): MatchDetail {
  return {
    match_id: matchId,
    game_creation: Date.now() - 86_400_000,
    game_duration: gameDuration,
    game_version: "14.24.1",
    queue_id: 420,
    queue_label: "Ranked Solo/Duo",
    map_id: 11,
    map_label: "Summoner's Rift",
    teams: buildTeams(viewerWin),
  };
}

const MOCK_MATCHES: Record<string, MatchDetail> = {
  EUW1_mock_1: createMatch("EUW1_mock_1", false, 25 * 60),
  EUW1_mock_2: createMatch("EUW1_mock_2", true, 30 * 60),
  EUW1_mock_3: createMatch("EUW1_mock_3", true, 40 * 60),
  EUW1_mock_4: createMatch("EUW1_mock_4", false, 20 * 60),
};

export const MOCK_MATCH_DETAIL = MOCK_MATCHES.EUW1_mock_1;

export function getMockMatchDetail(matchId: string): MatchDetail {
  return MOCK_MATCHES[matchId] ?? { ...MOCK_MATCH_DETAIL, match_id: matchId };
}

export { VIEWER_PUUID };

File: backend/app/services/match_detail.py
import json

from sqlalchemy.ext.asyncio import AsyncSession
from sqlmodel import select

from app.database.models import Champions, Matches, Participants
from app.schemas.match import (
    ChampionBanResponse,
    MatchDetailResponse,
    ObjectivesSummaryResponse,
    ParticipantDetailResponse,
    TeamDetailResponse,
)
from app.utils.game_labels import map_label, queue_label


async def user_has_match_access(
    session: AsyncSession, user_puuids: list[str], match_id: str
) -> bool:
    if not user_puuids:
        return False
    result = await session.execute(
        select(Participants.internal_id)
        .where(
            Participants.match_id == match_id,
            Participants.puuid.in_(user_puuids),
        )
        .limit(1)
    )
    return result.scalar_one_or_none() is not None


async def get_match_detail(
    session: AsyncSession, match_id: str, viewer_puuid: str | None
) -> MatchDetailResponse | None:
    result = await session.execute(
        select(Matches).where(Matches.match_id == match_id)
    )
    match = result.scalar_one_or_none()
    if not match or not match.detail_json:
        return None

    payload = json.loads(match.detail_json)
    teams_raw = payload.get("teams", [])

    ban_ids: set[int] = set()
    for team in teams_raw:
        ban_ids.update(team.get("bans", []))

    name_by_id: dict[int, str] = {}
    if ban_ids:
        champ_result = await session.execute(
            select(Champions.champion_id, Champions.name).where(
                Champions.champion_id.in_(ban_ids)
            )
        )
        name_by_id = dict(champ_result.all())

    teams: list[TeamDetailResponse] = []

    for team in teams_raw:
        participants: list[ParticipantDetailResponse] = []
        for p in team.get("participants", []):
            puuid = p.get("puuid", "")
            is_viewer = bool(
                viewer_puuid and puuid == viewer_puuid
            ) or p.get("is_viewer", False)
            participants.append(
                ParticipantDetailResponse(
                    puuid=puuid,
                    riot_id=p.get("riot_id"),
                    champion_id=p["champion_id"],
                    champion_name=p["champion_name"],
                    position=p["position"],
                    win=p["win"],
                    kills=p["kills"],
                    deaths=p["deaths"],
                    assists=p["assists"],
                    cs=p["cs"],
                    gold_earned=p["gold_earned"],
                    damage_to_champions=p["damage_to_champions"],
                    vision_score=p["vision_score"],
                    items=p.get("items", []),
                    summoner_spells=p.get("summoner_spells", []),
                    is_viewer=is_viewer,
                )
            )
        obj = team.get("objectives", {})
        bans = [
            ChampionBanResponse(
                champion_id=ban_id,
                champion_name=name_by_id.get(ban_id, f"Champion {ban_id}"),
            )
            for ban_id in team.get("bans", [])
        ]
        teams.append(
            TeamDetailResponse(
                team_id=team["team_id"],
                win=team["win"],
                bans=bans,
                objectives=ObjectivesSummaryResponse(
                    baron=obj.get("baron", 0),
                    dragon=obj.get("dragon", 0),
                    rift_herald=obj.get("rift_herald", 0),
                    tower=obj.get("tower", 0),
                    inhibitor=obj.get("inhibitor", 0),
                ),
                participants=participants,
            )
        )

    return MatchDetailResponse(
        match_id=match.match_id,
        game_creation=match.game_creation,
        game_duration=match.game_duration,
        game_version=match.game_version,
        queue_id=match.queue_id,
        queue_label=queue_label(match.queue_id),
        map_id=match.map_id,
        map_label=map_label(match.map_id),
        teams=teams,
    )
